crates/rpc/rpc-builder/src/rate_limiter.rs

```diff
@@ -0,0 +1,113 @@
+use tower::Layer;
+
+/// Rate limiter for the RPC server.
+///
+/// Rate limits expensive calls such as debug_ and trace_.
+#[derive(Debug, Clone)]
+pub(crate) struct RpcRequestRateLimiter {
+    inner: Arc<RpcRequestRateLimiterInner>,
+}
+
+impl RpcRequestRateLimiter {
+    pub(crate) fn new(rate_limit: usize) -> Self {
+        Self {
+            inner: Arc::new(RpcRequestRateLimiterInner {
+                call_guard: PollSemaphore::new(Arc::new(Semaphore::new(rate_limit))),
+            }),
+        }
+    }
+}
+
+impl<S> Layer<S> for RpcRequestRateLimiter {
+    type Service = RpcRequestRateLimitingService<S>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        RpcRequestRateLimitingService::new(inner, self.clone())
+    }
+}
+
+/// Rate Limiter for the RPC server
+#[derive(Debug, Clone)]
+struct RpcRequestRateLimiterInner {
+    /// Semaphore to rate limit calls
+    call_guard: PollSemaphore,
+}
+
+/// A [`RpcServiceT`] middleware that rate limits RPC calls to the server.
+///
+#[derive(Debug, Clone)]
+pub struct RpcRequestRateLimitingService<S> {
+    /// The rate limiter for RPC requests
+    rate_limiter: RpcRequestRateLimiter,
+    /// The inner service being wrapped
+    inner: S,
+}
+
+impl<S> RpcRequestRateLimitingService<S> {
+    pub(crate) fn new(service: S, rate_limiter: RpcRequestRateLimiter) -> Self {
+        Self { inner: service, rate_limiter }
+    }
+}
+
+impl<'a, S> RpcServiceT<'a> for RpcRequestRateLimitingService<S>
+where
+    S: RpcServiceT<'a> + Send + Sync + Clone + 'static,
+{
+    type Future = RateLimitingRequestFuture<S::Future>;
+
+    fn call(&self, req: Request<'a>) -> Self::Future {
+        let method_name = req.method_name();
+        if method_name.starts_with("trace_") || method_name.starts_with("debug_") {
+            RateLimitingRequestFuture {
+                fut: self.inner.call(req),
+                guard: Some(self.rate_limiter.inner.call_guard.clone()),
+            }
+        } else {
+            // if we don't need to rate limit, then there
+            // is no need to get a semaphore permit
+            RateLimitingRequestFuture {
+                fut: self.inner.call(req),
+                guard: None,
+            }
+        }
+    }
+}
+
+/// Response future.
+#[pin_project::pin_project]
+pub struct RateLimitingRequestFuture<F> {
+    #[pin]
+    fut: F,
+    guard: Option<PollSemaphore>,
+}
+
+impl<F> std::fmt::Debug for RateLimitingRequestFuture<F> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str("RateLimitingRequestFuture")
+    }
+}
+
+impl<F: Future<Output = MethodResponse>> Future for RateLimitingRequestFuture<F> {
+    type Output = F::Output;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        if let Some(mut guard) = self.guard.clone() {
+            match guard.poll_acquire(cx) {
+                Poll::Pending => return Poll::Pending,
+                _ => {},
+            }
```

Comment:
this will always acquire a new permit and immediately drop it.

what we can do here is add another Option for the acquired permit, and set it if the guard returns ready.
once the guard returns ready we can set the guard field to `None` this ensures we keep the acquired permit until we resolve the future:

```rust
this.permit = ready!(guard);
this.guard = None; 
```

Reply:
I'm running into some issues working with `Pin`.
Any idea how I can resolve this?

```rust

    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
        if let Some(mut guard) = self.guard.clone() {
            match guard.poll_acquire(cx) {
                Poll::Pending => return Poll::Pending,
                Poll::Ready(Some(permit)) => {
                    self.as_mut().permit = Some(permit);
```

Error:
```
cannot borrow data in dereference of `Pin<&mut RateLimitingRequestFuture<F>>` as mutable
trait `DerefMut` is required to modify through a dereference, but it is not implemented for `Pin<&mut RateLimitingRequestFuture<F>>`
```

Should I implement [DerefMut](https://doc.rust-lang.org/std/index.html)::[ops](https://doc.rust-lang.org/std/ops/index.html)::[DerefMut](https://doc.rust-lang.org/std/ops/trait.DerefMut.html#) or something?